batch/batch/front_end/front_end.py

```diff
@@ -1257,32 +1259,108 @@ async def ui_get_billing(request, userdata):
     return await render_template('batch', request, userdata, 'billing.html', page_context)
 
 
+async def _query_billing_projects(db, user=None, billing_project=None):
+    args = []
+
+    where_conditions = []
+
+    if user:
+        where_conditions.append("JSON_CONTAINS(users, '%s')")
+        args.append(user)
+
+    if billing_project:
+        where_conditions.append('billing_projects.name = %s')
+        args.append(billing_project)
+
+    if where_conditions:
+        where_condition = f'WHERE {" AND ".join(where_conditions)}'
+    else:
+        where_condition = ''
+
+    sql = f'''
+SELECT billing_projects.name as billing_project, users, msec_mcpu, `limit`, SUM(`usage` * rate) as cost
+FROM (
+  SELECT billing_project, JSON_ARRAYAGG(`user`) as users
+  FROM billing_project_users
+  GROUP BY billing_project
+) AS t
+RIGHT JOIN billing_projects
+  ON t.billing_project = billing_projects.name
```

Comment:
I don't understand how this works. `t` might have a record like:
```
hail, []
```
and RIGHT JOINing that with
```
hail, ...
```
will still make `users` be the empty array, right?

Reply:
No. t would not have that record at all because it's aggregating over the billing_project_users table. We want to get all possible billing projects even those without users. That's why you need the right join.